### PR TITLE
Fix Issue-171 by adding a validation to check for querystring paramet…

### LINKED ISSUE #171: BVN404 11.2.3 custom redirects with parameters not working in episerver 11.12
Hi Team,

We have upgraded episerver from 10.10 to 11.12 also upgrade BVN404 to 11.2.3. After the upgrade we migrated the custom redirect url from dds to sql. After performing this step we noticed that some custom url not working.

Issue is with the urls having parameters.
if entry is like below in custom redirect manager:

old url: /content/file
new url: /legacy

we are accessing with the url http://mysite.com/content/file?doc=filename

On BVN version 10.1 we were redirected to http://mysite.com/legacy?doc=filename

On BVN version 11.2.3 we are redirected to http://mysite.com/legacy/?doc=filename 
(somehow an extra '/' is added at the end of url after 'legacy')

Thank you in advance.

Looking forward to your response as we are totally stuck.

### PR DESCRIPTION
**Pull request by espierre**
Fix #171 
Fix Issue-171 by adding a validation to check for querystring parameters before adding the trailing slash

### CODE REVIEW

**Comment by marisks** [comment]
@espierre Could you please add tests for this?

**Comment by espierre** [comment]
@marisks Added tests to cover issue-171 changes.

File: tests/Geta.404Handler.Tests/CustomRedirectCollectionTests.cs [hunk 1]
```diff
@@ -276,5 +276,65 @@ public void Find_gives_nullreferenceexception_with_partial_deleted_hit()
...
+            var actual = collection.Find(urlToFind.ToUri());
+
+            Assert.Equal(expected, actual.NewUrl);
+        }
+
+        /// <summary>
+        /// https://github.com/Geta/404handler/issues/171
+        /// </summary>
+        [Fact]
+        public void Find_does_heep_slash_when_querystring_parameter()
```
Comment (marisks) [inline tests/Geta.404Handler.Tests/CustomRedirectCollectionTests.cs:304]:
```suggestion
        public void Find_does_keep_slash_when_querystring_parameter()
```

File: src/Geta.404Handler/Core/CustomRedirects/CustomRedirectCollection.cs [hunk 2]
```diff
@@ -144,7 +144,16 @@ string RemoveSlash(string s)
             }
 
             var redirCopy = new CustomRedirect(cr);
-            var newUrl = AppendSlash(redirCopy.NewUrl);
+            var newUrl = string.Empty;
```
Comment (marisks) [inline src/Geta.404Handler/Core/CustomRedirects/CustomRedirectCollection.cs:147]: You can use an expression `? :` together with the assignment.

Comment (espierre) [reply src/Geta.404Handler/Core/CustomRedirects/CustomRedirectCollection.cs]: Changed the condition operator to include in variable assignment.

**Comment by lanorkin** [comment]
@marisks here is a somewhat related PR from me with tests #168 which received no attention; I described some cases with incorrect behavior like this one in TODOs in tests in that PR

For this PR, there are more cases to be fixed - @espierre might be interesting for you, several TODOs starting from here https://github.com/Geta/404handler/pull/168/files#diff-c3448f2e875671b96f1597ec3936a893R139

**Comment by marisks** [comment]
@lanorkin The main issue with #168 is that it has a lot of TODOs. I understood that it is under development. As no improvements came to fix TODOs, I didn't look at it anymore.
Now it also has conflicts with the main branch too.

**Comment by lanorkin** [comment]
@marisks well idea was to write more tests to document current (well, current at that moment year ago) behavior; TODOs meant to highlight incorrect (but existing) behavior, which should have been fixed with future bugs & PRs like this one.

pure PR with tests, like it is said in title; I refer to it here, because it contains more cases that can be fixed in regard to trailing slash issue

**Comment by marisks** [comment]
@lanorkin I can't approve PR that has failing tests. That PR can be kept open for the devs to look at cases that can be fixed.

**Comment by lanorkin** [comment]
@marisks all the tests were passing at the moment; that's the idea of documenting behavior - you have passing tests for all the cases, even if that cases does not look OK from business perspective; TODOs added to highlight that though test is passing now, it is something which should be changed in future to reflect real business needs. 

let's switch to discussion in #168 if you need any details; @espierre sorry for discussing it here

File: src/Geta.404Handler/Core/CustomRedirects/CustomRedirectCollection.cs [hunk 3]
```diff
@@ -144,7 +144,7 @@ string RemoveSlash(string s)
             }
 
             var redirCopy = new CustomRedirect(cr);
-            var newUrl = AppendSlash(redirCopy.NewUrl);
+            var newUrl = (url.IndexOf("?", StringComparison.Ordinal) > 0) ? redirCopy.NewUrl : AppendSlash(redirCopy.NewUrl);
```
Comment (marisks) [inline src/Geta.404Handler/Core/CustomRedirects/CustomRedirectCollection.cs:147]: Please remove unnecessary brackets.
```suggestion
            var newUrl = url.IndexOf("?", StringComparison.Ordinal) > 0 ? redirCopy.NewUrl : AppendSlash(redirCopy.NewUrl);
```

**Comment by espierre** [comment]
I saw that this PR was merged int your code base.  Do yo know how long it would take to have it available in the Episerver nuget feed?

**Comment by valdisiljuconoks** [comment]
Sorry guys for delays (vacations). I'll make it happen so you can get official release.

Stay safe!

**Pull request merged**